### PR TITLE
Expressions in action payloads

### LINKED ISSUE #341: Expressions in action payloads
## Use case

The history of department D15 (example data from [OData-Temporal]) shall be merged into D08. Assume we already have a function `toTimeslices` that converts `Departments('D15')/history` into the `TimesliceWithPeriod` format so that
```java
GET ~/Departments('D15')/history/Temporal.toTimeslices()
```
returns
```json
[{"Timeslice": {
   "From": "2010-01-01",
   "To": "2011-01-01",
   "Name": "Services",
   "Budget": 1100
}},
 {"Timeslice": {
   "From": "2011-01-01",
   "To": "9999-12-31",
   "Name": "Services",
   "Budget": 1700
}}]
```
If action parameters were specified in the URL, one could write:
```java
POST ~/Departments('D08')/history/Temporal.Update(deltaTimeslices=@delta)
&@delta=$root/Departments('D15')/history/Temporal.toTimeslices()
```
But action parameters must be passed in the payload, and this does not allow aliases and expressions.

## Suggestion
```java
POST ~/Departments('D08')/history/Temporal.Update

{
 "deltaTimeslices@expressionUrl":
  "$root/Departments('D15')/history/Temporal.toTimeslices()"
}
```
Or, because in a structured payload, a structured expression fits better than an expression string that must be parsed, we borrow notation from [CSDL](https://docs.oasis-open.org/odata/odata-csdl-json/v4.01/odata-csdl-json-v4.01.html#sec_ApplyClientSideFunctions):
```java
POST ~/Departments('D08')/history/Temporal.Update

{
 "deltaTimeslices@expressionCsdl": {
  "$Function": "Temporal.toTimeslices",
  "$Apply": [
   {"$Path": "$root/Departments('D15')/history"}
  ]
 }
}
```

Imported from [ODATA-1505](https://issues.oasis-open.org/browse/ODATA-1505)

### PR DESCRIPTION
**Pull request by HeikoTheissen**
Fixes #341

### CODE REVIEW

File: odata-json-format/16 Bound Function.md [hunk 1]
```diff
@@ -234,6 +234,27 @@ properties, or just the [entity reference](#EntityReference), as
 appropriate to the action.
 Stream typed parameter values are represented following the same rules as inlined [stream properties](#StreamProperty).
 
+Alternatively, values of non-binding parameters MAY be specified as common expressions
+[OData-URL, section 5.1.1](#ODataURL) that the service evaluates relative to the
+action. They are then encoded as a name/value
```
Comment (HeikoTheissen) [inline odata-json-format/16 Bound Function.md:239]: Relative to what are expressions in function invocations evaluated?
```
/service/Employees(23)/CheckAbsence(Employee=@colleague,Date=@date)
?@colleague=Manager
&@date=now() add duration'P1D'
```

Comment (ralfhandl) [reply odata-json-format/16 Bound Function.md:239]: I'd say the base URL in this example is scheme plus host (and port) plus `/service/Employees(23)/` - note the trailing slash.

File: odata-json-format/16 Bound Function.md [hunk 2]
```diff
@@ -234,6 +234,31 @@ properties, or just the [entity reference](#EntityReference), as
...
+Example ##ex: An employee requests leave from their manager for the next two weeks.
+```json
+POST /service/Employees(23)/RequestLeave
+Host: host
+Content-Type: application/json
+
+{
+  "StartDate@expressionUrl": "now()",
+  "EndDate@expressionUrl": "now() add duration'P14D'",
+  "Approver@expressionUrl": "$root/Employees(23)/Manager"
```
Comment (ralfhandl) [inline odata-json-format/16 Bound Function.md:254]: We should add an example where relative resolution matters, and explain how it is resolved.

**Review by ralfhandl — COMMENTED** [review]
Going into the right direction

File: odata-json-format/16 Bound Function.md [hunk 3]
```diff
@@ -234,6 +234,48 @@ properties, or just the [entity reference](#EntityReference), as
 appropriate to the action.
 Stream typed parameter values are represented following the same rules as inlined [stream properties](#StreamProperty).
 
+Alternatively, values of non-binding parameters MAY be specified as common expressions
+[OData-URL, section 5.1.1](#ODataURL). In the case of a bound action
+these MAY contain path expressions [OData-URL, section 5.1.1.15](#ODataURL), which
+the service evaluates on the binding parameter. Such parameters are encoded as a name/value
+pair where the name is the name of the parameter followed by `@expressionUrl` and
```
Comment (ralfhandl) [inline odata-json-format/16 Bound Function.md:241]:
```suggestion
the service evaluates on the binding parameter value. Such parameters are encoded as name/value
pairs where the name is the name of the parameter followed by `@expressionUrl` and
```

File: odata-url-conventions/4 Resource Path.md [hunk 4]
```diff
@@ -569,7 +569,34 @@ syntax rule define the grammar for invoking functions, for example to help filte
 and order resources identified by the `resourcePath` of the URL.
 - The `aliasAndValue` syntax rule defines
 the grammar for providing function parameter values using Parameter
-Alias Syntax, see [OData-Protocol](#ODataProtocol).
+Alias Syntax, see [OData-Protocol](#ODataProtocol). The aliases can contain [common
+expressions](#CommonExpressionSyntax). In the case of a bound function
+these MAY contain [path expressions](#PathExpressions), which
+the service evaluates on the binding parameter.
```
Comment (ralfhandl) [inline odata-url-conventions/4 Resource Path.md:575]:
```suggestion
the service evaluates on the binding parameter value.
```

File: odata-url-conventions/4 Resource Path.md [hunk 5]
```diff
@@ -569,7 +569,34 @@ syntax rule define the grammar for invoking functions, for example to help filte
...
+Example ##ex: An employee's leave requests pending their manager's approval
+for the next two weeks:
+```
+http://host/service/Employees(23)/self.PendingLeaveRequests(StartDate=@start,
+  EndDate=@end,Approver=@approver)
+  ?@start=now()
+  &@end=now() add duration'P14D'
+  &@approver=Manager
+```
+The expression `Manager` is evaluated on the binding parameter `Employees(23)`.
```
Comment (ralfhandl) [inline odata-url-conventions/4 Resource Path.md:587]:
```suggestion
The expression `Manager` is evaluated on the binding parameter value `Employees(23)`.
```

File: odata-json-format/16 Bound Function.md [hunk 6]
```diff
@@ -234,6 +234,48 @@ properties, or just the [entity reference](#EntityReference), as
...
+Host: host
+Content-Type: application/json
+
+{
+  "StartDate@expressionUrl": "now()",
+  "EndDate@expressionUrl": "now() add duration'P14D'",
+  "Approver@expressionUrl": "Manager"
+}
+```
+The expression `Manager` is evaluated on the binding parameter `Employees(23)`.
```
Comment (ralfhandl) [inline odata-json-format/16 Bound Function.md:258]:
```suggestion
The expression `Manager` is evaluated on the binding parameter value `Employees(23)`.
```

**Review by ralfhandl — COMMENTED** [review]
+1, with minor change proposals

File: odata-json-format/16 Bound Function.md (hunk 1 above)
Comment (ralfhandl) [reply odata-json-format/16 Bound Function.md]: Seems to be covered now: anything I missed?

**Review by ralfhandl — DISMISSED** [review]

File: odata-json-format/16 Bound Function.md [hunk 7]
```diff
@@ -234,6 +234,49 @@ properties, or just the [entity reference](#EntityReference), as
...
+
+::: example
+Example ##ex: An employee requests leave from their manager for the next two weeks:
+```json
+POST /service/Employees(23)/self.RequestLeave
+Host: host
+Content-Type: application/json
+
+{
+  "StartDate@expressionUrl": "now()",
```
Comment (ralfhandl) [inline odata-json-format/16 Bound Function.md:254]:
```suggestion
  "StartDate@expression": "now()",
```
Rename throughout

File: odata-url-conventions/4 Resource Path.md
Comment (ralfhandl) [inline odata-url-conventions/4 Resource Path.md:1]: Please make this change a separate pull request

File: odata-json-format/1 Introduction.md [hunk 8]
```diff
@@ -25,8 +25,11 @@ Section | Feature / Change | Issue
 Fragment portion of Context URL is not percent-encoded| 
 [368](https://github.com/oasis-tcs/odata-specs/issues/368)
 [Section ##ControlInformationmediaodatamedia]| 
- `mediaContentType` can be `null`| 
+`mediaContentType` can be `null`| 
 [536](https://github.com/oasis-tcs/odata-specs/issues/536)
+[Section ##ActionInvocation]| 
+Allow common expressions in action payload|
```
Comment (ralfhandl) [inline odata-json-format/1 Introduction.md:31]:
```suggestion
Allow common expressions in action payloads| 
```

**Review by ralfhandl — DISMISSED** [review]

**Review by ralfhandl — DISMISSED** [review]

**Review by ralfhandl — APPROVED** [review]
